### ISSUE TITLE
[RFC] Support `prim.TupleConstruct` in `MaximizeValueSemantics` pass

### ISSUE THREAD

**Issue by makslevental** [llvm/torch-mlir#952]
**Motivation**: Models that return tuples that contain results of views or view-like ops; e.g., Huggingface's implementation of [OPT](https://github.com/huggingface/transformers/blob/522a9ece4baeb5abfec8953ef76469a530e987d5/src/transformers/models/opt/modeling_opt.py#L685-L700). If I had to guess/bet, I would guess/bet that most of Huggingface's transformer implementations do this kind of thing (in order to have composable layers). So really the motivation is to support Huggingface's implementations of various transformer models (which are probably valuable models for the ML community). 

**Proposal**: After some discussion with @ramiro050, the approach might look something like an abstract interpreter specialization (i.e., `abstractlyInterpretTupleConstructSlice`) that takes a sub-slice of the slice found by `matchAndRewrite`, which starts at a `prim.TupleConstruct` and ends at a `prim.TupleIndex` or a `return`. It is unclear whether that sub-slice should contain only sequences of `prim.TupleConstruct`s or not (depending on whethere there are torch ops that operate on a tuple of tensors and return a tuple of tensors - [there might in fact be such torch ops](https://pytorch.org/docs/stable/generated/torch.chunk.html)).

**MWE of fail to lower** (to `linalg`):

```mlir
module attributes {torch.debug_module_name = "MyModule"} {
  func.func @forward(%arg0: !torch.vtensor<[1,8,1024],f32>) -> !torch.tuple<tensor<[1,8,16,64],f32>, tensor<[1,8,16,64],f32>> {
    %int-1 = torch.constant.int -1
    %int1 = torch.constant.int 1
    %int16 = torch.constant.int 16
    %int64 = torch.constant.int 64
    %0 = torch.copy.to_tensor %arg0 : !torch.tensor<[1,8,1024],f32>
    %1 = torch.prim.ListConstruct %int1, %int-1, %int16, %int64 : (!torch.int, !torch.int, !torch.int, !torch.int) -> !torch.list<int>
    %2 = torch.aten.view %0, %1 : !torch.tensor<[1,8,1024],f32>, !torch.list<int> -> !torch.tensor<[1,8,16,64],f32>
    %3 = torch.prim.TupleConstruct %2, %2 : !torch.tensor<[1,8,16,64],f32>, !torch.tensor<[1,8,16,64],f32> -> !torch.tuple<tensor<[1,8,16,64],f32>, tensor<[1,8,16,64],f32>>
    return %3 : !torch.tuple<tensor<[1,8,16,64],f32>, tensor<[1,8,16,64],f32>>
  }
}
```

and fails with error

```mlir
  * Pattern (anonymous namespace)::AbstractlyInterpretCopyToNonValueTensorOpUsersWithinABlock : 'torch.copy.to_tensor -> ()' {
Trying to match "(anonymous namespace)::AbstractlyInterpretCopyToNonValueTensorOpUsersWithinABlock"
    ** Failure : unsupported op `torch.prim.TupleConstruct` encountered during abstract analysis
```

cc @powderluv

**Comment by silvasean** [comment]
I think the existing logic should be able to be generalized -- basically prim.TupleIndex'es of prim.TupleConstruct ops just create additional aliases.

**Comment by gpetters94** [comment]
@silvasean I'm picking this one up - how would that logic generalize to TupleConstruct chains? For example, in OPT the final return is of a prim.TupleConstruct, but the arguments of that are also tuples, three levels deep. Would each of the intermediate levels produce aliases too, or just the return op at the end of it?

**Comment by ramiro050** [comment]
> @silvasean I'm picking this one up - how would that logic generalize to TupleConstruct chains? For example, in OPT the final return is of a prim.TupleConstruct, but the arguments of that are also tuples, three levels deep. Would each of the intermediate levels produce aliases too, or just the return op at the end of it?

My understanding is that this was put on hold because there is a larger issue that needs to be solved first. Namely, none of the current backends to torch-mlir support returning nested tuples. So even if the value semantics issue is fixed, there would be no way of running the model.

**Comment by silvasean** [comment]
Correct. I suspect that the OPT model needs to be wrapped in Python to just return the logits or something. Usually these models return a bunch of information (various "scores" and such) but really what we are interested in to test correctness is just the raw logits or token ids.

You can see an example here for HF bert: https://github.com/iree-org/iree-torch/blob/09d8a98c3f6557de4aa1dd915a6f10d2b30597bc/examples/bert.py#L48

**Comment by makslevental** [comment]
>Correct. I suspect that the OPT model needs to be wrapped in Python to just return the logits or something.

The architecture of the model does not easily lend itself to that kind of refactoring; see [here](https://github.com/makslevental/SHARK/blob/c2cabbfa64aa4c1b7a5c0269056324ad2e0bcd1f/tank/pytorch/opt/README.md#bigger-problems).

**Comment by silvasean** [comment]
Can you just do `return self.forward()[0]` on the result of the model?

**Comment by makslevental** [comment]
> Can you just do `return self.forward()[0]` on the result of the model? Or `self.forward()[0][0]` or something.

```
    %852 = torch.prim.TupleConstruct %134, %139 : !torch.tensor<[1,12,7,64],f32>, !torch.tensor<[1,12,7,64],f32> -> !torch.tuple<tensor<[1,12,7,64],f32>, tensor<[1,12,7,64],f32>> loc(#loc0)
    %853 = torch.prim.TupleConstruct %199, %204 : !torch.tensor<[1,12,7,64],f32>, !torch.tensor<[1,12,7,64],f32> -> !torch.tuple<tensor<[1,12,7,64],f32>, tensor<[1,12,7,64],f32>> loc(#loc0)
    %854 = torch.prim.TupleConstruct %259, %264 : !torch.tensor<[1,12,7,64],f32>, !torch.tensor<[1,12,7,64],f32> -> !torch.tuple<tensor<[1,12,7,64],f32>, tensor<[1,12,7,64],f32>> loc(#loc0)
    %855 = torch.prim.TupleConstruct %319, %324 : !torch.tensor<[1,12,7,64],f32>, !torch.tensor<[1,12,7,64],f32> -> !torch.tuple<tensor<[1,12,7,64],f32>, tensor<[1,12,7,64],f32>> loc(#loc0)
    %856 = torch.prim.TupleConstruct %379, %384 : !torch.tensor<[1,12,7,64],f32>, !torch.tensor<[1,12,7,64],f32> -> !torch.tuple<tensor<[1,12,7,64],f32>, tensor<[1,12,7,64],f32>> loc(#loc0)
    %857 = torch.prim.TupleConstruct %439, %444 : !torch.tensor<[1,12,7,64],f32>, !torch.tensor<[1,12,7,64],f32> -> !torch.tuple<tensor<[1,12,7,64],f32>, tensor<[1,12,7,64],f32>> loc(#loc0)
    %858 = torch.prim.TupleConstruct %499, %504 : !torch.tensor<[1,12,7,64],f32>, !torch.tensor<[1,12,7,64],f32> -> !torch.tuple<tensor<[1,12,7,64],f32>, tensor<[1,12,7,64],f32>> loc(#loc0)
    %859 = torch.prim.TupleConstruct %559, %564 : !torch.tensor<[1,12,7,64],f32>, !torch.tensor<[1,12,7,64],f32> -> !torch.tuple<tensor<[1,12,7,64],f32>, tensor<[1,12,7,64],f32>> loc(#loc0)
    %860 = torch.prim.TupleConstruct %619, %624 : !torch.tensor<[1,12,7,64],f32>, !torch.tensor<[1,12,7,64],f32> -> !torch.tuple<tensor<[1,12,7,64],f32>, tensor<[1,12,7,64],f32>> loc(#loc0)
    %861 = torch.prim.TupleConstruct %679, %684 : !torch.tensor<[1,12,7,64],f32>, !torch.tensor<[1,12,7,64],f32> -> !torch.tuple<tensor<[1,12,7,64],f32>, tensor<[1,12,7,64],f32>> loc(#loc0)
    %862 = torch.prim.TupleConstruct %739, %744 : !torch.tensor<[1,12,7,64],f32>, !torch.tensor<[1,12,7,64],f32> -> !torch.tuple<tensor<[1,12,7,64],f32>, tensor<[1,12,7,64],f32>> loc(#loc0)
    %863 = torch.prim.TupleConstruct %799, %804 : !torch.tensor<[1,12,7,64],f32>, !torch.tensor<[1,12,7,64],f32> -> !torch.tuple<tensor<[1,12,7,64],f32>, tensor<[1,12,7,64],f32>> loc(#loc0)
    %864 = torch.prim.TupleConstruct %852, %853, %854, %855, %856, %857, %858, %859, %860, %861, %862, %863 ....
    %865 = torch.prim.TupleConstruct %851, %864 
```

so maybe you can do `self.forward()[0][0]...<25 timtes?>...[0]` but 

1. That seems "wrong"
2. I don't know how the nesting changes between runs (as a function of parameters? as a function of data?). Maybe someone that's knowledgeable about either this model or HF transformers in general can comment but it still wouldn't be a robust solution for *all* HF models

**Comment by silvasean** [comment]
I don't think it is nested that deep. I only see 3 levels of nesting in the IR snippet you presented. And the model should have a stable API in this regard somehow (might need some digging) -- otherwise people would not be able to use it at the Python level either.

**Comment by gpetters94** [comment]
If we're just looking for the tensor output then `self.forward()[0]` should work. I'm working on a refactor of the model to see if we can get that to pass.

**Comment by silvasean** [comment]
What is self.forward()[0] -- logits? token ids?

**Comment by gpetters94** [comment]
It's a tensor that is the "last hidden state" as per [this](https://huggingface.co/docs/transformers/model_doc/opt#transformers.OPTModel.forward).  I'm not familiar with how NLP models work, so I don't know if the other returns are something we should be supporting, but it's the only non-optional return.
